### ISSUE TITLE
Version after dependency changes?

### ISSUE THREAD

**Issue by tvthompsonii** [semver/semver#148]
I don't see anything in the spec that says what a packages version should be if it needed to be published solely because a dependency changed. For example, let's say my package is at version 1.2.3 and a dependent product just moved from its version 2.3.4 up to version 3.0.0. That version change signals that my product needs to be rebuilt and published since the dependecy has changed it's public api. 

What would my product now be? Is it up to what changes I had to make? I mean, if I changed my public api to get things working, then obviously I would release 2.0.0. But otherwise, would it be 1.2.4 or 1.3.0? 

Both seem a little "off". My package technically is the same "version" it was before. I would almost rather have a "build number" at the end. My package is just a rebuild, maybe no code changes, so it would go from 1.2.3-1 to 1.2.3-2, for example. Users still know they have the same "version" but there is still something in the version number to help the build system with its dependency management.

What is the "semvar way" though?

**Comment by remcoros** [comment]
So this just happened: https://github.com/quartznet/quartznet/issues/229, anyone more familiar with semver can comment on this?

**Comment by FichteFoll** [comment]
If the dependency whose major was updated is linked or bundled statically and does not impose changes for the project's API it should be okay to just increase patch or minor if you suspect few issues to arise.
If however you are linking or just working dynamically in general, I think increasing minor is a must and probably even major since a library can have two types of interfaces, incoming and outgoing. Upgrading the logger dependency would be an outgoing interface change and must be considered part of the public API.

**Comment by e2** [comment]
I agree with @FichteFoll here:

"If the dependency whose major was updated is linked or bundled statically and does not impose changes for the project's API it should be okay to just increase patch or minor if you suspect few issues to arise."

The trail of thought is: single responsibility principle.

A dependency should only be responsible for the API it exposes itself. 

(Even if it's API can be augmented by another dependency).

#### Example:

1. `Tool 1.0.0` depends on `Library 1.0.0`. 
2. Both are written in a scripting language interpreter `Interpreter 1.0.0`.
3. So, both `Tool 1.0.0` and `Library 1.0.0` declare dependency on `Interpreter ~> 1.0`
4. After many months, `Interpreter 1.0.0` is no longer supported and has outstanding bugs. These manifest in `Library 1.0.0` and are fixed in `Interpreter 1.1.0`.

So, `Library` bumps it's dependency on `Interpreter` to `Interpreter 1.1.0` to side-step the bug.

#### Question: 

What should the version of `Library` be?.

#### Answer: 

`Library 1.0.1` (patchlevel change).

#### Why? 

Because the API didn't change - specifically, there are no code changes necessary in `Tool` (or apps using `Tool`), so there's no need for even a minor bump in `Library`. 

Even if `Library's` API is augmented by a different interpreter version (`Interpreter 1.1.0`), it doesn't matter.

#### Details:

Although `Interpreter 1.1.0` can "extend" the API of `Library`, it isn't `Library's` responsibility to document and manage that API. That's because `Tool` can further bump the dependency on `Interpreter` to `Interpreter 1.2.0` - outside the control of `Library`.

#### In practice:

The interesting thing here is that users typically see `Interpreter` as a "special case", which is "outside" SemVer. So they expect `Library 1.0.0` to be bumped to `Library 2.0.0`, because the following case seems "broken":

1. User has `Interpreter 1.0.0`
2. User attempts to install `Tool 1.0.0` and the latest version of `Library` (`1.0.1`), expecting patch level to not have a significant impact.
3. Installation fails since `Library 1.0.1` cannot be used on `Interpreter 1.0.0`.
4. User freaks out and complains that SemVer has been violated.
5. Insanely long discussions ensure...

#### Suggestion:

So it might help if this case was accurately covered by SemVer. 

(That according to single responsibility rule, a change in patchlevel does not guarantee that dependencies will stay satiable).

I think it's an error to suggest a major bump in this case because: you no longer can patch the previous major version with automatic upgrades.

E.g. If `Library` was erroneously bumped to `2.0.0` (just to avoid the "install breakage" above), then  `Tool` is still using `Library 1.0.0`, which not only contains the bug, but needs a dependency change to to the major version of `Library 2.0.0` to be fixed. And that misleadingly suggests an API change. (Which never happened).

In short: a backward-compatible bugfix should be reflected by no more than a patchlevel change, regardless of the circumstances (dependencies or dependencies of dependencies needed for the patch).

I'm wondering how to present this in a generic way (without the notion of an 'interpreter').

**Comment by jwdonahue** [comment]
@tvthompsonii, any reason to keep this issue open?  If not, please close it at your earliest possible convenience.

**Comment by tvthompsonii** [comment]
This seems to be an open debate, and no clear answer has been given. My recommendation would still be to add an optional build number to the spec. If you re-release really is just a rebuild, not something hitting the corner cases raised above such as needing to install a new set of dependent libraries, then a build number extension would seem ideal.
Closing this as other threads have picked up this conversation (and I have been requested to close).

**Comment by MrMino** [comment]
What about dependency systems, where conflicting requirements cannot be resolved by consuming the new release of the dependency? E.g. Python:

1. **A** 1.0.0 depends on **B** 1.0.0
2. **B** updates to 2.0.0
3. **A** has to constrain itself to **B** < 2.0.0, so it updates its version to 1.0.1
4. New release of **A** requires **B** 2.0.0 to add a feature
5. There is a library **C** in the system that still depends on **B** 1.0.0.
6. New release of **A** would now break the system. Should the new release of **A** update its minor or should it update its major?

**Comment by epicfaace** [comment]
What about upgrading a peer dependency (in this case React) by a major version bump? What is the recommended semver bump for this case?

**Comment by ljharb** [comment]
that’s absolutely a breaking change, because it forces consumers to upgrade the peer dep;

**Comment by chrisknoll** [comment]
> that’s absolutely a breaking change, because it forces consumers to upgrade the peer dep.
> 
> npm peer deps are part of your api - adding one is breaking, as is bumping it at all - not just a major.

Are you sure about this? If you go from 3.0.1 to 3.2.2 in an npm package, don't you still have to do a npm install the same as if it was 3.0.1 to 4.0.2?  So because the peers are forced to upgrade to their peer dep (based on a minor version update) your logic says that is a breaking change. I'm not so sure about that.  

I'm currently faced with moving from springboot 1.5.3 to something in the 5.x line..and if I can keep all the existing application behavior and api the same going from 1.5.3 to 5.x, I'm thinking of doing this in a minor release.  I use major release changes to indicate breaking backwards compatibility (ie features go away or the data model changes in some significant way (scalars become collections, etc).  

I definitely would not call this a 'patch' release, simply because the scope of code changes goes beyond calling them a simple 'fix'.

But doing this move will result in zero documentation changes, zero api changes, zero behavior changes, so...that smells like a minor release to me. Even if there are new Spring framework APIs made available to the 5.x release, i think all those changes are internal-facing...

**Comment by ljharb** [comment]
@chrisknoll `npm install` doesn't install peer deps, and if you have a conflicting version installed, `npm ls` will start erroring, indicating your dependency graph is invalid. So yes, if you raise the minimum bar for a peer dep, ever, it is a breaking change. You can, however, *extend* it - ie, you can have a peer dep of `^1.5.3 || ^2 || ^3 || ^4 || ^5` - and this will just be semver-minor.

**Comment by in-op** [comment]
There is a very colorful discussion regarding the inclusion of the Rust compiler on a minor version bump for the cryptography python package here: https://github.com/pyca/cryptography/issues/5771

**Comment by ljharb** [comment]
@alex’s summary there fails to recognize that platform support requirements are part of the public API, and thus semver adherence would have avoided the issues in that thread.

**Comment by GeorgeWL** [comment]
a related question, what if the package has a minor update due to addignan optional new feature e.g. 1.2.3 to 1.3.0, but you as the consumer are not using any of the code from that new feature.

should one increase your semver by a patch? or leave it as no change?

**Comment by chrisknoll** [comment]
Semver isn't a 'how is it used?' statement, it is 'how has it changed'. New functions imply minor update so 1.23 to 1.3.0 is expected here.  If it was backwards compatibility breaking, it would go to 2.0.0.

**Comment by ljharb** [comment]
@GeorgeWL if you indeed don't care about any of the bug fixes or additions since 1.2.3, then there's no reason to upgrade.

**Comment by jwdonahue** [comment]
@GeorgeWL, what tool chain are you using? Are you purveying an app or a library?

> ... or leave it as no change?

No change, is never an option, if you're publishing an update. There should never be two different content publications with the same product name and version number.

As for the rest of your questions, it really depends on where that updated package lands. In most of the commonly used environments, the package lands somewhere that your customer will have access to it directly, independent of your usage. Many packaging schemes pile everything into one directory somewhere, and possibly also some form of cache. So, if you're publishing a library, that depends on that updated library, your bits and theirs wind up in the same scope as your customer's bits. If you're adding features to that pool, even ones you don't use, they may be features your customer must test or at least analyze the security surfaces of, before deploying. 

If you're publishing a stand-alone application, that doesn't mix into the rest of your customers bits, or pollute any caches, then you can probably get by with a minor bump. You should really be aware of how your particular development silo works, from end to end, and apply common sense. 

Without further context, no one blanket statement is the right answer to your question, but the minor bump for minor bump answer is mostly likely correct and rarely has down side.

Speaking of down sides, does your update include critical security fixes, independent of that updated dependency? If you want to ensure your bug fixes get taken up quickly, publish them without that library update first. Many teams have either automated ingestion or relaxed restrictions on taking up patches, but more constrained rules around new features and breaking changes.

**Comment by ljharb** [comment]
To clarify, the question was about when a _dependency_ publishes an update, do you need to update to it if you're not using anything in the update?

I'm sure it depends on the ecosystem, but in npm, semver ranges should mean that consumers will get the latest by default, but aren't _required to_ unless you've also updated.

**Comment by jwdonahue** [comment]
> a related question, what if the package has a minor update due to adding an optional new feature e.g. 1.2.3 to 1.3.0, but you as the consumer are not using any of the code from that new feature.

>should one increase your semver by a patch? or leave it as no change?

The question, as I read it, has to do with what kind of version bump should you take, if any, on your own bits, if you update a dependency. The "should one increase your semver..." part of that seems to imply that something which is about to be published, was updated, why else would you consider a version bump, if you're not about to publish something?

**Comment by chrisknoll** [comment]
If you update a dependency it should, at a minimum, be reflected in the patch version part of semver (major.minor.PATCH).   However, the new dependency may expose new configuration options that your app can support. This is new functionality, so it would be at least a minor update (major.MINOR.patch).  Sometimes, tho, a dependency may lead to a tool chain update (ie: this dependency requires JDK 17 which is different than the minimum supported JDK of 8) That type of toolchain update I'd consider major (MAJOR.minor.patch).

If you don't need or want the new dependency version, then you're not actually changing anything in your code, so no semver required because you're not releasing anything new.

**Comment by remcoros** [comment]
I settled on not thinking about dependencies at all anymore. If you consider all dependencies as your own internal code, it makes things really simple for both maintainers and end-users: breaking changes > Major, New features > Minor, Fixes > Patch. Doesn't matter if those come from "internal" changes or "external dependencies", just treat them the same.
